File: Packs/pai-knowledge-system/src/server/mcp-wrapper.ts
#!/usr/bin/env bun
/**
 * MCP Server Wrapper for PAI Knowledge System
 *
 * This script provides a simple command-line interface to the Graphiti MCP server.
 * It handles JSON-RPC communication and provides simple wrappers around the MCP tools.
 *
 * Usage:
 *   bun run src/server/mcp-wrapper.ts add_episode "Episode title" "Episode body"
 *   bun run src/server/mcp-wrapper.ts search_nodes "search query"
 *   bun run src/server/mcp-wrapper.ts get_status
 */

import { createMCPClient } from "./lib/mcp-client.js";
import { cli } from "./lib/cli.js";

/**
 * Command definitions
 */
interface Command {
  name: string;
  description: string;
  handler: (args: string[]) => Promise<{ success: boolean; data?: unknown; error?: string }>;
}

/**
 * MCP Wrapper class
 */
class MCPWrapper {
  private commands: Map<string, Command>;

  constructor() {
    this.commands = new Map();
    this.registerCommands();
  }

  /**
   * Register all available commands
   */
  private registerCommands(): void {
    this.addCommand({
      name: "add_episode",
      description: "Add knowledge to graph",
      handler: this.cmdAddEpisode.bind(this),
    });

    this.addCommand({
      name: "search_nodes",
      description: "Search entities",
      handler: this.cmdSearchNodes.bind(this),
    });

    this.addCommand({
      name: "search_facts",
      description: "Search relationships",
      handler: this.cmdSearchFacts.bind(this),
    });

    this.addCommand({
      name: "get_episodes",
      description: "Get recent episodes",
      handler: this.cmdGetEpisodes.bind(this),
    });

    this.addCommand({
      name: "get_status",
      description: "Get graph status",
      handler: this.cmdGetStatus.bind(this),
    });

    this.addCommand({
      name: "clear_graph",
      description: "Delete all knowledge",
      handler: this.cmdClearGraph.bind(this),
    });

    this.addCommand({
      name: "health",
      description: "Check server health",
      handler: this.cmdHealth.bind(this),
    });
  }

  /**
   * Add a command
   */
  private addCommand(command: Command): void {
    this.commands.set(command.name, command);
  }

  /**
   * Get a command by name
   */
  private getCommand(name: string): Command | undefined {
    return this.commands.get(name);
  }

  /**
   * List all available commands
   */
  listCommands(): Command[] {
    return Array.from(this.commands.values());
  }

  /**
   * Command: add_episode
   */
  private async cmdAddEpisode(args: string[]): Promise<{ success: boolean; data?: unknown; error?: string }> {
    if (args.length < 2) {
      return {
        success: false,
        error: "Usage: add_episode <title> <body> [source_description]",
      };
    }

    const title = args[0];
    const body = args[1];
    const sourceDescription = args[2];

    const client = createMCPClient();
    const result = await client.addEpisode({
      name: title,
      episode_body: body,
      source_description,
    });

    return result;
  }

  /**
   * Command: search_nodes
   */
  private async cmdSearchNodes(args: string[]): Promise<{ success: boolean; data?: unknown; error?: string }> {
    if (args.length < 1) {
      return {
        success: false,
        error: "Usage: search_nodes <query> [limit]",
      };
    }

    const query = args[0];
    const limit = args.length > 1 ? parseInt(args[1], 10) : 5;

    const client = createMCPClient();
    const result = await client.searchNodes({ query, limit });

    return result;
  }

  /**
   * Command: search_facts
   */
  private async cmdSearchFacts(args: string[]): Promise<{ success: boolean; data?: unknown; error?: string }> {
    if (args.length < 1) {
      return {
        success: false,
        error: "Usage: search_facts <query> [limit]",
      };
    }

    const query = args[0];
    const limit = args.length > 1 ? parseInt(args[1], 10) : 5;

    const client = createMCPClient();
    const result = await client.searchFacts({ query, limit });

    return result;
  }

  /**
   * Command: get_episodes
   */
  private async cmdGetEpisodes(args: string[]): Promise<{ success: boolean; data?: unknown; error?: string }> {
    const limit = args.length > 0 ? parseInt(args[0], 10) : 5;

    const client = createMCPClient();
    const result = await client.getEpisodes({ limit });

    return result;
  }

  /**
   * Command: get_status
   */
  private async cmdGetStatus(): Promise<{ success: boolean; data?: unknown; error?: string }> {
    const client = createMCPClient();
    const result = await client.getStatus();

    return result;
  }

  /**
   * Command: clear_graph
   */
  private async cmdClearGraph(args: string[]): Promise<{ success: boolean; data?: unknown; error?: string }> {
    // Safety check
    if (!args.includes("--force")) {
      return {
        success: false,
        error: "This will delete ALL knowledge. Use --force to confirm.",
      };
    }

    const client = createMCPClient();
    const result = await client.clearGraph();

    return result;
  }

  /**
   * Command: health
   */
  private async cmdHealth(): Promise<{ success: boolean; data?: unknown; error?: string }> {
    const client = createMCPClient();
    const result = await client.testConnection();

    return result;
  }

  /**
   * Print help message
   */
  printHelp(): void {
    cli.blank();
    cli.header("PAI Knowledge System - MCP Wrapper", 60);
    cli.blank();
    cli.info("Simple command-line interface to the Graphiti MCP server.");
    cli.blank();
    cli.info("Usage:");
    cli.dim("  bun run src/server/mcp-wrapper.ts <command> [args...]");
    cli.blank();
    cli.info("Commands:");
    cli.blank();

    const commands = this.listCommands();
    const maxLength = Math.max(...commands.map((c) => c.name.length));

    commands.forEach((cmd) => {
      const paddedName = cmd.name.padEnd(maxLength);
      cli.dim(`  ${paddedName}  ${cmd.description}`);
    });

    cli.blank();
    cli.info("Examples:");
    cli.blank();
    cli.dim('  # Add knowledge to graph');
    cli.dim('  bun run src/server/mcp-wrapper.ts add_episode "Test Episode" "This is a test episode"');
    cli.blank();
    cli.dim('  # Search for entities');
    cli.dim('  bun run src/server/mcp-wrapper.ts search_nodes "PAI" 10');
    cli.blank();
    cli.dim('  # Get graph status');
    cli.dim('  bun run src/server/mcp-wrapper.ts get_status');
    cli.blank();
    cli.dim('  # Check server health');
    cli.dim('  bun run src/server/mcp-wrapper.ts health');
    cli.blank();
  }

  /**
   * Execute a command
   */
  async execute(commandName: string, args: string[]): Promise<number> {
    const command = this.getCommand(commandName);

    if (!command) {
      cli.error(`Unknown command: ${commandName}`);
      cli.blank();
      cli.info(`Available commands: ${Array.from(this.commands.keys()).join(", ")}`);
      return 1;
    }

    // Execute command
    const result = await command.handler(args);

    // Output result
    if (result.success) {
      if (result.data !== undefined) {
        console.log(JSON.stringify(result.data, null, 2));
      }
      return 0;
    } else {
      cli.error(`Error: ${result.error}`);
      return 1;
    }
  }
}

/**
 * Main function
 */
async function main() {
  const args = process.argv.slice(2);

  // Show help if no arguments
  if (args.length === 0) {
    const wrapper = new MCPWrapper();
    wrapper.printHelp();
    process.exit(0);
  }

  // Check for help flag
  if (args[0] === "-h" || args[0] === "--help") {
    const wrapper = new MCPWrapper();
    wrapper.printHelp();
    process.exit(0);
  }

  // Execute command
  const commandName = args[0];
  const commandArgs = args.slice(1);

  const wrapper = new MCPWrapper();
  const exitCode = await wrapper.execute(commandName, commandArgs);
  process.exit(exitCode);
}

// Run main function
main().catch((error) => {
  cli.error("Unexpected error:");
  console.error(error);
  process.exit(1);
});

File: Packs/madeinoz67-knowledge-system/src/server/mcp-wrapper.py
#!/usr/bin/env python3
"""
MCP Server Wrapper for PAI Knowledge System

This script provides a simple command-line interface to the Graphiti MCP server.
It handles SSE (Server-Sent Events) communication and provides simple wrappers
around the MCP tools.

Usage:
    python mcp-wrapper.py add_episode "Episode title" "Episode body"
    python mcp-wrapper.py search_nodes "search query"
    python mcp-wrapper.py get_status
"""

import sys
import json
import urllib.request
import urllib.parse
from typing import Dict, Any, Optional

# MCP Server Configuration
MCP_SERVER_URL = "http://localhost:8000/sse"
TIMEOUT = 30  # seconds


def call_mcp_tool(tool_name: str, arguments: Dict[str, Any]) -> Dict[str, Any]:
    """
    Call an MCP tool on the Graphiti server.

    Args:
        tool_name: Name of the MCP tool to call
        arguments: Dictionary of arguments for the tool

    Returns:
        Response from the MCP server
    """
    # Create JSON-RPC request
    payload = {
        "jsonrpc": "2.0",
        "id": 1,
        "method": "tools/call",
        "params": {
            "name": tool_name,
            "arguments": arguments
        }
    }

    try:
        # For SSE transport, we need to use POST to the endpoint
        # The Graphiti server handles both JSON-RPC and SSE
        req = urllib.request.Request(
            MCP_SERVER_URL,
            data=json.dumps(payload).encode('utf-8'),
            headers={
                'Content-Type': 'application/json',
            }
        )

        with urllib.request.urlopen(req, timeout=TIMEOUT) as response:
            response_data = json.loads(response.read().decode('utf-8'))

            if 'error' in response_data:
                return {
                    'success': False,
                    'error': response_data['error']
                }

            return {
                'success': True,
                'data': response_data.get('result', {})
            }

    except urllib.error.HTTPError as e:
        return {
            'success': False,
            'error': f'HTTP {e.code}: {e.reason}'
        }
    except urllib.error.URLError as e:
        return {
            'success': False,
            'error': f'Connection error: {e.reason}'
        }
    except Exception as e:
        return {
            'success': False,
            'error': str(e)
        }


def cmd_add_episode(args):
    """Add an episode to the knowledge graph."""
    if len(args) < 2:
        return {
            'success': False,
            'error': 'Usage: add_episode <title> <body> [source_description]'
        }

    title = args[0]
    body = args[1]
    source_description = args[2] if len(args) > 2 else None

    arguments = {
        'name': title,
        'episode_body': body,
    }

    if source_description:
        arguments['source_description'] = source_description

    return call_mcp_tool('add_episode', arguments)


def cmd_search_nodes(args):
    """Search for nodes in the knowledge graph."""
    if len(args) < 1:
        return {
            'success': False,
            'error': 'Usage: search_nodes <query> [limit]'
        }

    query = args[0]
    limit = int(args[1]) if len(args) > 1 else 5

    arguments = {
        'query': query,
        'limit': limit
    }

    return call_mcp_tool('search_nodes', arguments)


def cmd_search_facts(args):
    """Search for facts/relationships in the knowledge graph."""
    if len(args) < 1:
        return {
            'success': False,
            'error': 'Usage: search_facts <query> [limit]'
        }

    query = args[0]
    limit = int(args[1]) if len(args) > 1 else 5

    arguments = {
        'query': query,
        'limit': limit
    }

    return call_mcp_tool('search_facts', arguments)


def cmd_get_episodes(args):
    """Get recent episodes from the knowledge graph."""
    limit = int(args[0]) if len(args) > 0 else 5

    arguments = {
        'limit': limit
    }

    return call_mcp_tool('get_episodes', arguments)


def cmd_get_status(args):
    """Get the status of the knowledge graph."""
    return call_mcp_tool('get_status', {})


def cmd_clear_graph(args):
    """Clear all data from the knowledge graph."""
    # Safety check
    if '--force' not in args:
        return {
            'success': False,
            'error': 'This will delete ALL knowledge. Use --force to confirm.'
        }

    return call_mcp_tool('clear_graph', {})


def main():
    if len(sys.argv) < 2:
        print("Usage: python mcp-wrapper.py <command> [args...]")
        print("")
        print("Commands:")
        print("  add_episode <title> <body> [source]     Add knowledge to graph")
        print("  search_nodes <query> [limit]             Search entities")
        print("  search_facts <query> [limit]             Search relationships")
        print("  get_episodes [limit]                     Get recent episodes")
        print("  get_status                               Get graph status")
        print("  clear_graph --force                      Delete all knowledge")
        sys.exit(1)

    command = sys.argv[1]
    args = sys.argv[2:]

    # Map commands to functions
    commands = {
        'add_episode': cmd_add_episode,
        'search_nodes': cmd_search_nodes,
        'search_facts': cmd_search_facts,
        'get_episodes': cmd_get_episodes,
        'get_status': cmd_get_status,
        'clear_graph': cmd_clear_graph,
    }

    if command not in commands:
        print(f"Unknown command: {command}")
        print(f"Available commands: {', '.join(commands.keys())}")
        sys.exit(1)

    # Execute command
    result = commands[command](args)

    # Output result
    if result.get('success'):
        print(json.dumps(result['data'], indent=2))
        sys.exit(0)
    else:
        print(f"Error: {result.get('error')}", file=sys.stderr)
        sys.exit(1)


if __name__ == '__main__':
    main()
